sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/directconnectivity/ConsistencyWriter.java

```diff
@@ -318,34 +318,25 @@ boolean isGlobalStrongRequest(RxDocumentServiceRequest request, StoreResponse re
                     //only refresh on first barrier call, set to false for subsequent attempts.
                     barrierRequest.requestContext.forceRefreshAddressCache = false;
 
-                    //trace on last retry.
+                    //get max global committed lsn from current batch of responses, then update if greater than max of all batches.
                     if (writeBarrierRetryCount.getAndDecrement() == 0) {
                         logger.debug("ConsistencyWriter: WaitForWriteBarrierAsync - Last barrier multi-region strong. Responses: {}",
                             responses.stream().map(StoreResult::toString).collect(Collectors.joining("; ")));
+                        logger.debug("ConsistencyWriter: Highest global committed lsn received for write barrier call is {}", maxGlobalCommittedLsnReceived);
+                        return Mono.just(Boolean.FALSE);
                     }
 
                     return Mono.empty();
                     }).flux();
-        }).repeatWhen(s -> {
-            if (writeBarrierRetryCount.get() == 0) {
-                    return Flux.empty();
+        }).repeatWhen(s -> s.flatMap(x -> {
+            // repeat with a delay
+            if ((ConsistencyWriter.MAX_NUMBER_OF_WRITE_BARRIER_READ_RETRIES - writeBarrierRetryCount.get()) > ConsistencyWriter.MAX_SHORT_BARRIER_RETRIES_FOR_MULTI_REGION) {
+                return Flux.just(0L).delayElements(Duration.ofMillis(ConsistencyWriter.DELAY_BETWEEN_WRITE_BARRIER_CALLS_IN_MS));
```

Comment:
This works but you have some dummy value (`0L`) which makes it a bit hard to read.

please replace
`Flux.just(0L).delayElements(duration)` with `Mono.delay(duration).flux()`.

Reply:
Fixed